### PR TITLE
More efficient pletFields with list of fields

### PR DESCRIPTION
**Pull request by t1lde**
Addresses #182

Still need to make sure this is back-compatible, there's some polykind weirdness I'm unsure of

### CODE REVIEW

**Comment by t1lde** [comment]
(waiting on local build so I can fix after the merge)

File: Plutarch/DataRepr/Internal/Field.hs
Comment (t1lde) [inline Plutarch/DataRepr/Internal/Field.hs:242]: There's now *a lot* of type-family evaluation going on... with the nesting, and list-recursion the type errors are pretty crazy

File: Plutarch/DataRepr/Internal/HList.hs
Comment (t1lde) [inline Plutarch/DataRepr/Internal/HList.hs:113]: latest version of `base` has polykinded `Min` and `Max` that could be used if not for backcompat

File: Plutarch/DataRepr/Internal/Field.hs
Comment (L-as) [inline Plutarch/DataRepr/Internal/Field.hs:118]: This doesn't make sense. Why is `z` bound?

**Review by L-as — CHANGES_REQUESTED** [review]

File: Plutarch/DataRepr/Internal/Field.hs
Comment (t1lde) [reply Plutarch/DataRepr/Internal/Field.hs:118]: currently the field binding doesn't do any skipping, you're right that we can do better

Comment (t1lde) [reply Plutarch/DataRepr/Internal/Field.hs:118]: I'll try and get this working

Comment (L-as) [reply Plutarch/DataRepr/Internal/Field.hs:118]: BTW it's OK if the implementation doesn't make use of the fact that you only need some of the fields, since we can just fix that later in a non-breaking change.

File: examples/Examples/Field.hs
Comment (L-as) [inline examples/Examples/Field.hs:307]: I just noticed that there is an inefficiency that can't be solved easily, but we can make it better. The `force headList x` part should be plet'ed.

Comment (L-as) [reply examples/Examples/Field.hs:307]: It's still not perfect, because if you use it only once it should be inlined. There's not really a good way of doing that, but it's better that it be a plet IMO. It's very unintuitive for the field access to be expensive IMO. Maybe there's some other way we can make users `plet` it themselves.

**Review by L-as — APPROVED** [review]
Looks great, do you think this is solvable? Will merge since it doesn't affect the API.

**Comment by L-as** [comment]
>  Could you please add hrecField to Plutarch.Prelude now that pletFields is encouraged?

it's encouraged now!

File: examples/Examples/Field.hs
Comment (TotallyNotChase) [reply examples/Examples/Field.hs:307]: This could be somewhat problematic since, very often - people use a field only once. In which case, this would be a rather unnecessary `plet`. But I absolutely understand the concern.

Comment (TotallyNotChase) [reply examples/Examples/Field.hs:307]: On a related note: what's up with so much builtin duplication in that compiled output? Is that missing hoisting?

Comment (t1lde) [reply examples/Examples/Field.hs:307]: It would be a simple change to `plet` bind all the `force headList x` terms by default. The question is whether we want that.

I think it's acceptable enough for users to have to know to plet bind the terms if they're going to use them more than once, aside from changes to hoisting I'm not sure if we can have it work automagically

Comment (L-as) [reply examples/Examples/Field.hs:307]: We are in fact missing some `phoistAcyclic` here! Good catch. #177

Comment (L-as) [reply examples/Examples/Field.hs:307]: This is very odd, I can't find what's not hoisted.

Comment (L-as) [reply examples/Examples/Field.hs:307]: #195

Comment (TotallyNotChase) [reply examples/Examples/Field.hs:307]: It's ok - I'll be testing these parts pretty extensively in the coming days, hopefully we'll be able to catch everything one by one.

**Pull request merged**
